src/openms/source/FILTERING/DATAREDUCTION/Deisotoper.cpp

```diff
@@ -38,9 +38,292 @@
+  }
+
+  for (size_t current_peak = 0; current_peak != old_spectrum.size(); ++current_peak)
+  {
+    // only process peaks which are not already in a cluster. Would form clusters identical to the cluster they are assigned to,
+    // excluding its first peak(s), since peaks with lower mz are not considered for cluster-formation.
+    if (features[current_peak] != -1) continue;
+
+    // Monoisotopic peaks with intensity 0.0 interfere with averagine check when normalizing the spectrum peaks height
+    if (use_averagine_model && old_spectrum[current_peak].getIntensity() == 0.0) continue;
+
+    const double current_mz = old_spectrum[current_peak].getMZ();
+    if (add_up_intensity)
+    {
+      mono_iso_peak_intensity[current_peak] = old_spectrum[current_peak].getIntensity();
+    }
+    clusters.clear();
+    charges_of_extensions.clear();
+    for (int q = max_charge; q >= min_charge; --q) // important: test charge hypothesis from high to low
+    {      
+      // try to extend isotopes from mono-isotopic peak
+      // if appropriate extension larger than min_isopeaks possible:
+      //   - save charge q in mono_isotopic_peak[]
+      //   - annotate_charge all isotopic peaks with feature number
+
+      bool has_min_isopeaks = true;
+      const double tolerance_dalton = fragment_unit_ppm ? Math::ppmToMass(fragment_tolerance, current_mz) : fragment_tolerance;
+
+      // do not bother testing charges q (and masses m) with: m/q > precursor_mass/q (or m > precursor_mass)
+      if (has_precursor_data)
+      {
+        double current_theo_mass = (current_mz * q) - (Constants::PROTON_MASS * q);
+        if (current_theo_mass > (precursor_mass + tolerance_dalton))
+        {
+          continue;
+        }
+      }
+
+      extensions.clear();
+      extensions.push_back(current_peak);
+
+      float total_weight = old_spectrum[current_peak].getIntensity() * (old_spectrum[current_peak].getMZ() * q - (q - 1) * Constants::PROTON_MASS_U);
+      float total_intensity = old_spectrum[current_peak].getIntensity();
+      for (unsigned int i = 1; i < max_isopeaks; ++i)
+      {
+        const double expected_mz = current_mz + static_cast<double>(i) * Constants::C13C12_MASSDIFF_U / static_cast<double>(q);
+        const int p = old_spectrum.findNearest(expected_mz, tolerance_dalton);
+        if (p == -1)// test for missing peak
+        {
+          has_min_isopeaks = (i >= min_isopeaks);
+          break;
+        }
+        else
+        {
+          // compare to averagine distribution
+          if (use_averagine_model)
+          {
+            // compute average weight (weighted by intensity)
+            total_weight += old_spectrum[p].getIntensity() * (old_spectrum[p].getMZ() * q - (q - 1) * Constants::PROTON_MASS_U);
+            total_intensity += old_spectrum[p].getIntensity();
+
+            // generate averagine distribution
+            CoarseIsotopePatternGenerator gen(extensions.size() + 1);
```

Comment:
why only up to extensions.size() + 1?
do probabilities need to be normalized or not?

Reply:
> why only up to extensions.size() + 1?
The peaks in the theoretical distribution without a counterpart in the cluster would all result in an addition to the KL of 0 since P(x) would be 0, so we do not need to generate those.

> do probabilities need to be normalized or not?
The cluster we want to test and the theoretical distribution should sum up to the same amount, or else the KL does not make sense, as it compares distributions. Does specification of max_isotope lead to an unnormalized distribution? In that case renormalization would be necessary

Reply:
I am not sure just guessing if something could go wrong here.
Wouldn't it make sense to generate the full isotope pattern for the current mass and then compare the distributions?
Right now it (at least seems to) calculates it for multiple extensions.

Reply:
Can you check the documentation and code of the generator? Maybe be need to extend it if it is not clear from the docs.
My concern is that if we only generate e.g. 1st+2nd and 1st+2nd+3rd isotopic peaks. These might contribute to e.g., 80% and 95% of the full isotope pattern. I could imagine that it might make a huge difference normalized vs. non-normalized.

Reply:
Yes that makes sense, will do